algorithms/kernel/dtrees/forest/oneapi/cl_kernels/df_tree_level_build_helper_kernels.cl

```diff
@@ -0,0 +1,383 @@
+        const int nNodeProp  = NODE_PROPS; // num of split attributes for node
+        const int leafMark   = -1;
+        const int local_id   = get_sub_group_local_id();
+        const int nodeId     = get_global_id(1);
+        const int local_size = get_sub_group_size();
+
+        __global const int * node = nodeList + nodeId * nNodeProp;
+        const int offset          = node[0];
+        const int nRows           = node[1];
+        const int featId          = node[2];
+        const int splitVal        = node[3];
+        const int nRowsLeft       = node[4]; // num of items in Left part
+
+        if (featId != leafMark) // split node
+        {
+            int sum = 0;
+
+            for (int i = local_id; i < nRows; i += local_size)
+            {
+                int id                        = treeOrder[offset + i];
+                int toRight                   = (int)(data[id * nFeatures + featId] > splitVal);
+                int boundary                  = sum + sub_group_scan_exclusive_add(toRight);
+                int posNew                    = (toRight ? nRowsLeft + boundary : i - boundary);
+                treeOrderBuf[offset + posNew] = id;
+                sum += sub_group_reduce_add(toRight);
+            }
+        }
+    }
+
+    __kernel void getNumOfSplitNodes(const __global int * nodeList, int nNodes, __global int * nSplitNodes) {
+        const int local_id   = get_sub_group_local_id();
+        const int local_size = get_sub_group_size();
+        const int nNodeProp  = NODE_PROPS; // num of node properties in node
+        const int badVal     = -1;
+
+        int sum = 0;
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            sum += (int)(nodeList[i * nNodeProp + 2] != badVal);
+        }
+
+        sum = sub_group_reduce_add(sum);
+
+        if (local_id == 0)
+        {
+            nSplitNodes[0] = sum;
+        }
+    }
+
+    __kernel void convertSplitToLeaf(__global int * nodeList) {
+        const int nNodeProp = NODE_PROPS; // num of split attributes for node
+        const int leafMark  = -1;
+        const int id        = get_global_id(0);
+
+        nodeList[id * nNodeProp + 2] = leafMark;
+        nodeList[id * nNodeProp + 3] = leafMark;
+    }
+
+    __kernel void doNodesSplit(const __global int * nodeList, int nNodes, __global int * nodeListNew) {
+        const int nNodeProp  = NODE_PROPS; // num of split attributes for node
+        const int badVal     = -1;
+        const int local_id   = get_sub_group_local_id();
+        const int local_size = get_sub_group_size();
+
+        int nCreatedNodes = 0;
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            int splitNode      = (int)(nodeList[i * nNodeProp + 2] != badVal); // featId != -1
+            int newLeftNodePos = nCreatedNodes + sub_group_scan_exclusive_add(splitNode) * 2;
+            if (splitNode)
+            {
```

Comment:
Might make sense to convert it to a set of ternary operations if it is a hotspot

Reply:
Yes will consider it during further optimization